src/backend/columnar/columnar_metadata.c

```diff
@@ -860,34 +837,17 @@ ReserveStripe(Relation rel, uint64 sizeBytes,
 		metapage = InitMetapage(rel);
 	}
 
-	GetHighestUsedAddressAndId(metapage->storageId, &currLogicalHigh, &highestId);
-	SmgrAddr currSmgrHigh = logical_to_smgr(currLogicalHigh);
-
-	SmgrAddr resSmgrStart = next_block_start(currSmgrHigh);
-	uint64 resLogicalStart = smgr_to_logical(resSmgrStart);
-
-	uint64 resLogicalEnd = resLogicalStart + sizeBytes - 1;
-	SmgrAddr resSmgrEnd = logical_to_smgr(resLogicalEnd);
-
-	RelationOpenSmgr(rel);
-	uint64 nblocks = smgrnblocks(rel->rd_smgr, MAIN_FORKNUM);
-
-	while (resSmgrEnd.blockno >= nblocks)
-	{
-		Buffer newBuffer = ReadBuffer(rel, P_NEW);
-		ReleaseBuffer(newBuffer);
-		nblocks = smgrnblocks(rel->rd_smgr, MAIN_FORKNUM);
-	}
-
-	RelationCloseSmgr(rel);
+	uint64 first_row;
+	uint64 stripeId = ColumnarMetapageReserveStripe(rel, 0, &first_row);
```

Comment:
Regarding (3): do you have an opinion on whether to reserve the stripe ID along with the row numbers, or with the data offsets?

I don't think it matters much because stripe IDs are not a limited resource, so I figured it looked slightly nicer to reserve the stripe ID along with the row numbers.

Reply:
>I don't think it matters much because stripe IDs are not a limited resource, so I figured it looked slightly nicer to reserve the stripe ID along with the row numbers.

I agree with that, we can reserve them together with `reserved_row_number`

Reply:
We can't move the reservation of offsets earlier than flush, because we don't know how much space we need until then.

And I'll leave the other reservations a part of flush as well, for this PR. I think this PR is big enough  already :-(